### PR TITLE
Add KeyVault reference feature for VariableGroups Resource 

### PR DESCRIPTION
**Pull request by EliiseS**
## All Submissions:
-------------------------------------
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [NO] I have updated the documentation accordingly.
* [YES] I have added tests to cover my changes.
* [NA] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [NA] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open PRs for the same update/change?

## What is the current behavior?
-------------------------------------
Issue Number: #172 

## Does this introduce a breaking change?
-------------------------------------
- NO

## Other information
-------------------------------------
Currently the progress on this PR is blocked by a bug in the AzDO API GO SDK. For more information see the issue linked.

### CODE REVIEW

File: azuredevops/resource_variable_group.go [hunk 1]
```diff
@@ -55,7 +56,9 @@ func resourceVariableGroup() *schema.Resource {
 				Default:  false,
 			},
 			"variable": {
-				Type: schema.TypeSet,
+				Type:     schema.TypeSet,
+				Required: true,
+				MinItems: 1,
```
Comment (tmeckel) [inline azuredevops/resource_variable_group.go:61]: We need a `ConflictsWith = []string { "key-vault" }` attribute here, because you cannot have variables simultaneously with a KeyVault connection (well according to the web interface you can't set use variables with a key vault connection). And there should be a unit test, which validates that the provider won't accept both properties `variables` and `key_vault` set simultaneously.
AzDo KeyVault

File: azuredevops/resource_variable_group.go [hunk 2]
```diff
@@ -74,16 +77,37 @@ func resourceVariableGroup() *schema.Resource {
 						},
 					},
 				},
-				Required: true,
-				MinItems: 1,
+			},
+			"key_vault": {
+				Type:     schema.TypeSet,
```
Comment (tmeckel) [inline azuredevops/resource_variable_group.go:82]: We need a `ConflictsWith = []string { "variable" }` attribute here.

File: azuredevops/resource_variable_group.go [hunk 3]
```diff
@@ -248,10 +277,30 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
 	variableGroup := &taskagent.VariableGroupParameters{
 		Name:        converter.String(d.Get("name").(string)),
 		Description: converter.String(d.Get("description").(string)),
-		Variables:   &variableMap,
+		Variables:   variableMap,
+	}
+
+	key_vault := d.Get("key_vault").(*schema.Set).List()
+
```
Comment (tmeckel) [inline azuredevops/resource_variable_group.go:284]: Because `key-vault` is optional I'd recommend to use a `GetOk` and check if the the property is defined and continue.

File: azuredevops/resource_variable_group.go [hunk 4]
```diff
@@ -231,14 +259,15 @@ func deleteVariableGroup(clients *config.AggregatedClient, project *string, vari
 }
 
 // Convert internal Terraform data structure to an AzDO data structure
-func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableGroupParameters, *string) {
+func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableGroupParameters, *string, error) {
 	projectID := converter.String(d.Get("project_id").(string))
 	variables := d.Get("variable").(*schema.Set).List()
```
Comment (tmeckel) [inline azuredevops/resource_variable_group.go:264]: Because `variable` is optional I'd recommend to use a `GetOk` and check if the the property is defined and continue.

File: azuredevops/resource_variable_group.go [hunk 5]
```diff
@@ -74,16 +77,37 @@ func resourceVariableGroup() *schema.Resource {
...
+			},
+			"key_vault": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
```
Comment (tmeckel) [inline azuredevops/resource_variable_group.go:89]: A validation function is required here. I'd recommend `NoEmptyStrings` from the `validate` package
https://github.com/microsoft/terraform-provider-azuredevops/blob/534e23dfdc77b65c2ca8bd164e44b07a3a14441b/azuredevops/utils/validate/strings.go#L9

File: azuredevops/resource_variable_group.go (hunk 1 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go:61]: That's not entirely true, it does accept variables, because you have to specify which variables from key vault you'd like to have.

A separate problem from this is the fact that you can also set any keys and values you want for the key vault API and it'll work anyway.

File: azuredevops/resource_variable_group.go (hunk 2 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go:82]: As above, you need to specify the keys you'd like to use from your key vault

File: azuredevops/resource_variable_group.go (hunk 1 above)
Comment (tmeckel) [reply azuredevops/resource_variable_group.go:61]: I think the Product Group implemented the warning by purpose and not by accident. If you can use variables and simultaneously a KeyVault connection without receiving an error from the API it will not guarantee that it'll will really work. I'd recommend to get in touch with the product group and ask about why the webinterface shows a warning when you activate a keyvault connection. I have the suspect that the variables are still stored in the Variable Group but they won't be evaluated any more. That's why you receive no error while using the REST API.

File: azuredevops/resource_variable_group.go (hunk 2 above)
Comment (tmeckel) [reply azuredevops/resource_variable_group.go:82]: Ahhh okay ... got it ...

File: azuredevops/resource_variable_group.go [hunk 6]
```diff
@@ -55,35 +56,61 @@ func resourceVariableGroup() *schema.Resource {
...
 							Required: true,
 						},
 						"value": {
-							Type:     schema.TypeString,
-							Optional: true,
-							Default:  "",
+							Type:          schema.TypeString,
+							Optional:      true,
+							Default:       "",
+							ConflictsWith: []string{"key_vault"},
```
Comment (tmeckel) [inline azuredevops/resource_variable_group.go:72]: @EliiseS I saw that you used 'conflicts_with' after all. What made you change your mind?

Comment (EliiseS) [reply azuredevops/resource_variable_group.go:72]: I set it to conflict with is_secret and value, since you only need to specify the key name to get it from key vault. Being able to set others is misleading since they don't do anything.

File: azuredevops/utils/testhelper/hcl.go [hunk 7]
```diff
@@ -276,6 +276,27 @@ resource "azuredevops_variable_group" "vg" {
 	return fmt.Sprintf("%s\n%s", projectResource, variableGroupResource)
 }
 
+// TestAccVariableGroupResourceNoSecrets Similar to TestAccVariableGroupResource, but without a secret variable
+func TestAccVariableGroupResourceKeyVault(projectName string, variableGroupName string, allowAccess bool, keyVaultName, serviceEndpointId string) string {
+	variableGroupResource := fmt.Sprintf(`
```
Comment (tmeckel) [inline azuredevops/utils/testhelper/hcl.go:281]: HCL for acceptance tests are located in `azuredevops/utils/testhelper/hcl.go` provided as functions, so that you can easily reuse HCL code throughout various acceptance tests.

Comment (EliiseS) [reply azuredevops/utils/testhelper/hcl.go:281]: I don't really understand what you would like me to change? I can't seem to see any changes you are suggesting

Comment (tmeckel) [reply azuredevops/utils/testhelper/hcl.go:281]: All HCL code for the acceptances are located inside the `azuredevops/utils/testhelper/hcl.go` code file and encapsulated as functions with parameters like 

```go
// TestAccProjectResource HCL describing an AzDO project
func TestAccProjectResource(projectName string) string {
	if projectName == "" {
		panic("Parameter: projectName cannot be empty")
	}
	return fmt.Sprintf(`
resource "azuredevops_project" "project" {
	project_name       = "%s"
	description        = "%s-description"
	visibility         = "private"
	version_control    = "Git"
	work_item_template = "Agile"
}`, projectName, projectName)
}
```

This approach has the advantage that you could reuse the HCL by calling the function and getting a formatted HCL string back.

For example the function `TestAccProjectResource` is used at the following positions:

```
azuredevops/data_project_test.go:
  27: 	projectResource := testhelper.TestAccProjectResource(projectName)

azuredevops/resource_project_test.go:
  305: 				Config: testhelper.TestAccProjectResource(projectNameFirst),
  316: 				Config: testhelper.TestAccProjectResource(projectNameSecond),
  475: 				Config: testhelper.TestAccProjectResourceWithFeature(projectName, "disabled", "disabled"),
  489: 				Config: testhelper.TestAccProjectResourceWithFeature(projectName, "enabled", "disabled"),

azuredevops/resource_serviceendpoint_bitbucket_test.go:
  204: 	projectResource := testhelper.TestAccProjectResource(projectName)
  216: 	projectResource := testhelper.TestAccProjectResource(projectName)
  228: 	projectResource := testhelper.TestAccProjectResource(projectName)

```

My advice is to move the `fmt.Sprintf` statement, which creates the HCL for acceptance test, to a new function inside  `azuredevops/utils/testhelper/hcl.go`  and call this function in your `TestAccVariableGroupResourceKeyVault`  function.

File: GNUmakefile [hunk 8]
```diff
@@ -50,6 +50,10 @@ test-compile:
 	fi
 	go test -c $(TEST) $(TESTARGS)
 
+install:
```
Comment (nmiodice) [inline GNUmakefile:53]: If we are going to add this change here, we should update the instructions to make this the *documented* way to install locally. Then, every command a developer should need to run will be through `make`

File: azuredevops/resource_variable_group.go [hunk 9]
```diff
@@ -58,42 +65,85 @@ func resourceVariableGroup() *schema.Resource {
...
 					},
 				},
-				Required: true,
-				MinItems: 1,
 			},
 		},
 	}
 }
 
+const AzureKeyVaultType = "AzureKeyVault"
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go:128]: Do we need to export this from the package? If not, it should be `azureKeyVaultType`

File: azuredevops/resource_variable_group.go [hunk 10]
```diff
@@ -130,7 +180,15 @@ func resourceVariableGroupRead(d *schema.ResourceData, m interface{}) error {
...
 	}
 
-	flattenVariableGroup(d, variableGroup, &projectID)
+	err = flattenVariableGroup(d, variableGroup, &projectID)
+
+	if err != nil {
+		return fmt.Errorf("Error flattening variable group: %v", err)
+	}
+
+	if err != nil {
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go:189]: Duplicate error check here. The first seems to have the correct message

File: azuredevops/resource_variable_group.go [hunk 11]
```diff
@@ -130,7 +180,15 @@ func resourceVariableGroupRead(d *schema.ResourceData, m interface{}) error {
 		return fmt.Errorf("Error looking up variable group given ID (%v) and project ID (%v): %v", variableGroupID, projectID, err)
 	}
 
-	flattenVariableGroup(d, variableGroup, &projectID)
+	err = flattenVariableGroup(d, variableGroup, &projectID)
+
+	if err != nil {
+		return fmt.Errorf("Error flattening variable group: %v", err)
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go:186]: This may be flagged as a code smell by Sonar because the error text is duplicated

File: azuredevops/resource_variable_group.go [hunk 12]
```diff
@@ -155,7 +213,11 @@ func resourceVariableGroupRead(d *schema.ResourceData, m interface{}) error {
 
 func resourceVariableGroupUpdate(d *schema.ResourceData, m interface{}) error {
 	clients := m.(*config.AggregatedClient)
-	variableGroupParams, projectID := expandVariableGroupParameters(d)
+
+	variableGroupParams, projectID, err := expandVariableGroupParameters(d)
+	if err != nil {
+		return fmt.Errorf("Error expanding variable group resource data: %+v", err)
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go:219]: Similar note about the expand error text -- perhaps we can make this a constant?

File: azuredevops/resource_variable_group.go [hunk 13]
```diff
@@ -254,51 +320,108 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
 		Variables:   &variableMap,
 	}
 
-	return variableGroup, projectID
+	key_vault := d.Get(vgKeyVault).(*schema.Set).List()
+
+	// Note: If configured, this will be of length 1 based on the schema definition above.
+	// TODO: add unit test to cover if it's more than 1
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go]: Calling out the TODO, in case it was missed

File: azuredevops/resource_variable_group.go [hunk 14]
```diff
@@ -254,51 +320,108 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
...
+
+		if converter.ToBool(result.IsSecret, false) {
 			variable = tfhelper.FindMapInSetWithGivenKeyValue(d, vgVariable, vgName, varName)
 		} else {
 			variable = map[string]interface{}{
-				vgName:     varName,
-				vgValue:    converter.ToString(varVal.Value, ""),
-				vgIsSecret: false,
+				vgName:        varName,
+				vgValue:       converter.ToString(result.Value, ""),
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go]: Do we want to set this for KV groups? Does this cause a `non empty plan` scenario, because the configuration will always have these set to `""` for KV groups?

File: azuredevops/resource_variable_group_acceptance_test.go [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+				ResourceName:      tfVarGroupNode,
+				ImportStateIdFunc: testAccImportStateIDFunc(tfVarGroupNode),
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccVariableGroupKeyVault_CreateAndUpdate(t *testing.T) {
```
Comment (nmiodice) [inline azuredevops/resource_variable_group_acceptance_test.go:72]: Does this test use a real keyvault? Does the API not care if the keyvault actually exists?

File: go.mod [hunk 16]
```diff
@@ -9,10 +9,12 @@ require (
 	github.com/hashicorp/go-uuid v1.0.1
 	github.com/hashicorp/terraform v0.12.23
 	github.com/hashicorp/terraform-plugin-sdk v1.8.0
-	github.com/microsoft/azure-devops-go-api/azuredevops v1.0.0-b2
-	github.com/stretchr/testify v1.3.0
+	github.com/microsoft/azure-devops-go-api/azuredevops v1.0.0-b3
+	github.com/mitchellh/mapstructure v1.1.2
```
Comment (nmiodice) [inline go.mod]: Lots of changes here - can you provide context on why these are needed?

File: azuredevops/resource_variable_group.go (hunk 3 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go:322]: Right now this doesn't throw an exception when set like this, would you suggest to change the check to be `GetOK` from `len(key_vault) == 1`

File: azuredevops/resource_variable_group_acceptance_test.go (hunk 15 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group_acceptance_test.go:72]: They API does not care if a real key vault exists.

Comment (nmiodice) [reply azuredevops/resource_variable_group_acceptance_test.go:72]: Ugh, annoying! but it does make this test easier...

File: azuredevops/resource_variable_group.go (hunk 4 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go]: Variable is not optional

File: azuredevops/utils/testhelper/hcl.go (hunk 7 above)
Comment (EliiseS) [reply azuredevops/utils/testhelper/hcl.go]: Done

File: azuredevops/resource_variable_group.go [hunk 17]
```diff
@@ -58,42 +65,85 @@ func resourceVariableGroup() *schema.Resource {
...
+						},
+						vgExpires: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+			vgKeyVault: {
+				Type:          schema.TypeSet,
```
[the remark this answers is not in the document]
Comment (EliiseS) [reply azuredevops/resource_variable_group.go]: Changed

File: azuredevops/resource_variable_group.go (hunk 9 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go]: Awesome catch! Made a mental note to change it, but I lost the note :)

File: azuredevops/resource_variable_group.go (hunk 11 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go:184]: Should we not nest error messages? It's done everywhere

Comment (EliiseS) [reply azuredevops/resource_variable_group.go:184]: Aaa, you want in it a constant? I understand now, got it.

File: azuredevops/resource_variable_group.go (hunk 13 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go]: Right, removed because it's covered by the schema check so a unit test would be redudant in the end

File: azuredevops/resource_variable_group_acceptance_test.go (hunk 15 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group_acceptance_test.go:72]: I was honestly quite happy when I found out, because testing this was going to be such a pain otherwise!

File: GNUmakefile (hunk 8 above)
Comment (EliiseS) [reply GNUmakefile:53]: I've added them to the docs, let me know what you think

File: go.mod (hunk 16 above)
Comment (EliiseS) [reply go.mod]: I ran `go mod tidy` and `go mod vendor`, this was the result

File: GNUmakefile [hunk 18]
```diff
@@ -51,8 +51,7 @@ test-compile:
 	go test -c $(TEST) $(TESTARGS)
 
 install:
-	./scripts/build.sh 
-	./scripts/local-install.sh
+	./scripts/build.sh --SkipTests --Install
```
Comment (nmiodice) [inline GNUmakefile:58]: Suggest to remove `--SkipTests`

File: azuredevops/resource_variable_group.go [hunk 19]
```diff
@@ -254,51 +316,133 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
...
-				vgValue:    converter.ToString(varVal.Value, ""),
+				vgValue:    converter.ToString(variable.Value, ""),
 				vgIsSecret: false,
 			}
+
+			if variable.Expires != nil {
+				variableMap[vgExpires] = variable.Expires.String()
+			}
+			if variable.Enabled != nil {
+				variableMap[vgEnabled] = converter.ToBool(variable.Enabled, false)
```
Comment (nmiodice) [inline azuredevops/resource_variable_group.go:406]: `converter.ToBool(variable.Enabled, false)` will default to `false` even if `variable.Enabled` is `nil`, so the `nil` check is not needed.

Similar for `variable.ContentType` and also `variable.Expires`, if `converter.ToString(variable.Expires, "")` is used

File: azuredevops/resource_variable_group_test.go [hunk 20]
```diff
@@ -1,232 +1,116 @@
...
-	Id:         converter.String("100"),
-}
 
-/**
- * Begin unit tests
- */
 // verifies that the flatten/expand round trip yields the same build definition
-func TestAzureDevOpsVariableGroup_ExpandFlatten_Roundtrip(t *testing.T) {
+func TestVariableGroup_ExpandFlatten_Roundtrip(t *testing.T) {
+	// This definition matches the overall structure of what a configured git repository would
```
Comment (nmiodice) [inline azuredevops/resource_variable_group_test.go:23]: This comment appears to be unrelated to the implementation. There is also an errant comment related to build definitions in here

File: azuredevops/resource_variable_group_test.go [hunk 21]
```diff
@@ -1,232 +1,116 @@
-// +build all resource_variable_group
```
Comment (nmiodice) [inline azuredevops/resource_variable_group_test.go:1]: The diff is difficult to read in this PR based on the changes so I'll make a single comment up at the top:
- The test is difficult to follow because each tests appears to be testing (1) expand/flatten logic for the variable group, and (2) expand/flatten logic for the resource authorization. This is difficult to read, and belongs in a different test all together in my opinion.

- The expand/flatten for the variable group is verbose because you are testing the equality of specific fields. I understand the underlying issue is related to the typing challenges we discussed.

IMO, it would be much simpler to structure the test like so:

```
variableGroup := ... // init var group as you already do
flattenVariableGroup(d, variableGroup)
variableGroupExpanded = expandVariableGroup(d)
```

Then, assert the equality of `variableGroup` and `variableGroupExpanded` after converting both to JSON. This fixes all the typing issues, and should make the testing more verbose to any future changes to the resource implementation.

File: go.mod (hunk 16 above)
Comment (nmiodice) [reply go.mod]: Are all of the changes a result of the upgrade to the Go SDK, or did you upgrade other dependencies as well?

File: azuredevops/resource_variable_group.go (hunk 3 above)
Comment (nmiodice) [reply azuredevops/resource_variable_group.go:320]: If the default behavior is for `d.Get("key_vault").(*schema.Set).List()` to evaluate to an empty list if `key_vault` is not set, then I'm OK keeping it as is. If we need to protect against nil pointers then we should use `GetOk`

File: go.mod (hunk 16 above)
Comment (EliiseS) [reply go.mod]: Not that I'm aware of

File: azuredevops/resource_variable_group.go (hunk 14 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go]: Refactored

File: azuredevops/resource_variable_group.go (hunk 19 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go]: Fixed for everthing but expires, which is time

File: GNUmakefile (hunk 18 above)
Comment (nmiodice) [reply GNUmakefile:58]: Ignore this comment :) you have convinced me otherwise!

File: azuredevops/resource_variable_group_test.go (hunk 21 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group_test.go:1]: Great points, I've gone and given the file a new face. Hope you like it :)

File: azuredevops/resource_variable_group_test.go (hunk 20 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group_test.go]: Removed

File: azuredevops/resource_variable_group.go (hunk 3 above)
Comment (EliiseS) [reply azuredevops/resource_variable_group.go:320]: TestVariableGroup_ExpandFlatten_Roundtrip should have caught it if it was going to throw an nil pointer error, so I'll resolve this comment for now. If you can think you a test case that would fail, I'll be more than happy to add it and modify the code :)
